EIPS/eip-7002.md

```diff
@@ -0,0 +1,380 @@
+    return trie.root
+
+assert block.header.exits_root == compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+2. The list of exit operations contained in the block body **MUST** be equivalent to list of exits at the head of the exit precompile's exit message queue up to the maximum of `MAX_EXITS_PER_BLOCK`, respecting the order in the queue. This validation **MUST** be run after all transactions in the current block are processed and **MUST** be run before pre-block precompile storage calculations (i.e. a call to `update_exit_precompile()`) are performed. To illustrate:
+
+```python
+class Exit(object):
+    source_address: Bytes20
+    validator_pubkey: Bytes48
+
+queue_head_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)
+queue_tail_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT)
+num_exits_in_queue = queue_tail_index - queue_head_index
+num_exits_to_dequeue = min(num_exits_in_queue, MAX_EXITS_PER_BLOCK)
+
+# Retrieve exits from the queue
+expected_exits = []
+for i in range(num_exits_to_dequeue):
+    queue_storage_slot = EXIT_MESSAGE_QUEUE_STORAGE_OFFSET + (queue_head_index + i) * 2
+    source_address = address(SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot)[0:20]
+    validator_pubkey = (
+        SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1) + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)
+        + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1) + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)[0:16]
+    )
+    exit = Exit(
+        source_address=Bytes20(source_address),
+        validator_pubkey=Bytes48(validator_pubkey),
+    )
+    expected_exits.append(exit)
+
+# Compare retrieved exits to the list in the block body
+assert block.body.exits == expected_exits
+```
+
+A block that does not satisfy the above conditions **MUST** be deemed invalid.
+
+#### Block processing
+
+##### Per-block precompile storage calculations
+
+At the end of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. after processing all transactions and after performing the block body exit validations):
+
+* The exit precompile's exit queue is updated based on exits dequeued and the exit queue head/tail are reset if the queue has been cleared (`update_exit_queue()`)
+* The exit precompile’s excess exits are updated based on usage in the current block (`update_excess_exits()`)
+* The exit precompile's exit count is reset to 0 (`reset_exit_count()`)
+
+Specifically, the functionality is defined in pseudocode as the function `update_exit_precompile()`:
+
+```python
+###################
+# Public function #
+###################
+
+def update_exit_precompile():
+    update_exit_queue()
+    update_excess_exits()
+    reset_exit_count()
+    
+###########
+# Helpers #
+###########
+
+def update_exit_queue():
+    queue_head_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)
```

Comment:
readability could be improved across this document w/ helpers like

```python
def read_exit_precompile(TARGET_SLOT):
    return sload(EXIT_PRECOMPILE_ADDRESS, TARGET_SLOT)
```

then `read_exit_precompile(EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)`

and similar helpers for writing